transports/quic/src/transport.rs

```diff
@@ -76,48 +78,76 @@ pub enum Error {
     /// Error after the remote has been reached.
     #[error("{0}")]
     Established(Libp2pQuicConnectionError),
-    /// Error while working with IfWatcher.
+
     #[error("{0}")]
-    IfWatcher(std::io::Error),
+    Io(#[from] std::io::Error),
+
+    #[error("Background task crashed.")]
+    TaskCrashed,
 }
 
 impl Transport for QuicTransport {
     type Output = (PeerId, QuicMuxer);
     type Error = Error;
-    // type Listener = Pin<
-    //     Box<dyn Stream<Item = Result<ListenerEvent<Upgrade, Self::Error>, Self::Error>> + Send>,
-    // >;
-    type Listener = Self;
     type ListenerUpgrade = Upgrade;
     type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
 
-    fn listen_on(
-        &mut self,
-        addr: Multiaddr,
-    ) -> Result<Self::Listener, TransportError<Self::Error>> {
-        multiaddr_to_socketaddr(&addr)
-            .ok_or_else(|| TransportError::MultiaddrNotSupported(addr))?;
-        Ok(self.clone())
+    fn listen_on(&mut self, addr: Multiaddr) -> Result<ListenerId, TransportError<Self::Error>> {
+        let socket_addr =
+            multiaddr_to_socketaddr(&addr).ok_or(TransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new(listener_id, socket_addr, self.config.clone())
+            .map_err(TransportError::Other)?;
+        self.listeners.push(listener);
+        // Drop reference to dialer endpoint so that the endpoint is dropped once the last
+        // connection that uses it is closed.
+        // New outbound connections will use a bidirectional (listener) endpoint.
+        let _ = self.dialer.take();
+        Ok(listener_id)
     }
 
-    fn address_translation(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
+            listener.close(Ok(()));
```

Comment:
We can call ::close multiple times on one listener. I see there is a warning `Listener was already closed.` in this case, but who actually removes a listener from the list of listeners?

Reply:
The [`SelectAll`](https://docs.rs/futures/latest/futures/stream/select_all/struct.SelectAll.html#) type that we use to drive our listener streams removes a stream once it returns `Poll::Ready(None)`.
When the `Listener::report_closed` it set to `Some(Option<_>)` we return the inner Option. On the first iteration this will be a `TransportEvent`, but since we take that out of the option, the next time the listener is polled it will return `Poll::Ready(None)`: 
https://github.com/kpp/rust-libp2p/blob/4f7aed80cd2799b64931ba7512406fb48f7f266b/transports/quic/src/transport.rs#L297-L302

But just to be sure, I'll add a test to validate that this is actually the case.